Refactor CLI code

cmleinz opened the pull request:
# Summary

This is a draft PR for addressing #240. So far I've made one change to the CLI main function. I'm looking for feedback as to if this is a good path forward for the refactor, if it is I can proceed with making changes to the rest of the actions in a similar way.

## Architectural Changes

Adjust the CLI code to be more generic and reduce duplication

ChristopherRabotin commented on anise-cli/src/args.rs:
```diff
@@ -1,6 +1,6 @@
 use std::path::PathBuf;
 
-use clap::{Parser, Subcommand};
+use clap::{Args as ArgsT, Parser, Subcommand};
```
I typically prefer to rename the repo's structs to not conflict with imports, instead of the other way around. So in this case, you could rename the `pub struct Args` on line 8 to something else (`CliArgs`?), and still import clap's `Args` as is.

ChristopherRabotin commented on anise-cli/src/args.rs:
```diff
@@ -66,3 +55,17 @@ pub enum Actions {
 }
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, ArgsT)]
+pub(crate) struct TruncateById {
```
Could you add an example of how the command changes now that `TruncateById` derives clap's `Args`?

ChristopherRabotin commented on anise-cli/src/main.rs:
```diff
@@ -50,7 +53,9 @@ pub enum CliErrors {
     if var(LOG_VAR).is_err() {
-        set_var(LOG_VAR, "INFO");
+        unsafe {
+            set_var(LOG_VAR, "INFO");
```
Is `set_var` unsafe now?

cmleinz replied:
Yes it's been marked unsafeZ in the 2024 edition of rust

ChristopherRabotin commented on anise-cli/src/main.rs:
```diff
@@ -331,3 +249,56 @@ fn main() -> Result<(), CliErrors> {
     }
 }
+
+fn truncate_daf_by_id<R>(
```
Great idea!

ChristopherRabotin commented on anise/src/naif/daf/mod.rs:
```diff
@@ -45,6 +45,7 @@ pub trait NAIFRecord: AsBytes + FromBytes + Sized + Default + Debug {
 
 pub trait NAIFSummaryRecord: NAIFRecord + Copy {
     fn start_index(&self) -> usize;
+    fn data_type_i(&self) -> i32;
```
Instead of adding the `data_type_i`, I recommend making the `data_type()` function of [SPKSummaryRecord](https://docs.rs/anise/latest/anise/naif/spk/summary/struct.SPKSummaryRecord.html#method.data_type) and [BPCSummaryRecord](https://docs.rs/anise/latest/src/anise/naif/pck/mod.rs.html#39-44) generic.

The way you could do this is by adding a new type parameter to the trait as follows:

```rust
us core::error::Error;

pub trait NAIFSummaryRecord: NAIFRecord + Copy {
    type ErrorType: Error;
    /// (...)
    fn data_type(&self) -> Result<DafDataType, Self::ErrorType>;
}
```

Then in the BPCSummaryType, move the data_type function from this struct's own implementation to the struct's implementation of the NAIFSummaryRecord trait.

So around line 160 of naif/spk/summary.rs, you'd need to add the `type Error` definition.

```rust
impl NAIFSummaryRecord for SPKSummaryRecord {
    const NAME: &'static str = "SPKSummaryRecord";
    type ErrorType = EphemerisError;
    // (...)
   fn data_type(&self) -> Result<DafDataType, EphemerisError> {
        DafDataType::try_from(self.data_type_i).map_err(|source| EphemerisError::SPK {
            action: "converting data type from i32",
            source,
        })
    }
```

Repeat that for the BPC Summary structure.

Then in the CLI code, you can just use the data_type() function of the NAIFSummary directly, and remove the need to export the integer representing the data_type_i or even rebuilding the data type enum from that integer in the code.

Let me know if you'd like more clarifications here.

ChristopherRabotin commented:
Thanks for your work, @cmleinz ! This is a great start, I've just made a recommendation on the change to NAIFSummary which I think will make it more generic.

cmleinz commented:
I addressed the comments thus far, still haven't made the rest of the changes though. Let me know what you think! As for your comment on how the CLI will change with Args, I believe it will remain the same. I compared the outputs of the help commands for trunc-daf-by-id between this branch and master and couldn't find any differences.

Additionally for another CLI I maintain I derive Args and I'm pretty sure everything automagically works the same, I'll look a little deeper though to be sure

ChristopherRabotin commented on anise-cli/src/main.rs:
```diff
@@ -331,3 +253,58 @@ fn main() -> Result<(), CliErrors> {
+
+fn truncate_daf_by_id<R>(
+    args::TruncateById {
+        input,
```
I've not seen this syntax before. Does this unpack the struct on call? I read about this idea, but I didn't know that it was stabilized.

cmleinz replied:
Essentially yes, it's pattern I like for this sort of thing where the function consumes it's argument and the struct is defined primarily as a helper struct, all of whose fields you intend to use within the function's body. It's pretty common in the axum crate!

ChristopherRabotin commented on anise-cli/src/main.rs:
```diff
@@ -331,3 +253,58 @@ fn main() -> Result<(), CliErrors> {
+                        }
+                    );
+
+    let segment = fmt.nth_data::<Type2ChebyshevSet>(idx).unwrap();
```
These unwrap calls should probably bubble up.

cmleinz replied:
Yep sounds good I can add to the error list

ChristopherRabotin commented:
Great work! I especially like the changes to `file2heap` and `NAIFPrettyPrint`. which I hadn't considered. The failing tests are because there is an ephem that's behind a Github secret that only the maintainers can use in CI.
